libs/crypto/keys_manager_impl.cpp

```diff
@@ -43,21 +43,31 @@ namespace iroha {
   }
 
   KeysManagerImpl::KeysManagerImpl(std::string account_name)
-      : account_name_(std::move(account_name)) {}
+      : account_name_(std::move(account_name)),
+        log_(logger::log("KeysManagerImpl")) {}
 
   bool KeysManagerImpl::validate(const iroha::keypair_t &keypair) const {
     std::string test = "12345";
     auto sig = iroha::sign(
         iroha::sha3_256(test).to_string(), keypair.pubkey, keypair.privkey);
-    return iroha::verify(
-        iroha::sha3_256(test).to_string(), keypair.pubkey, sig);
+    if (not iroha::verify(
+        iroha::sha3_256(test).to_string(), keypair.pubkey, sig)) {
+      log_->error("key validation failed");
+      return false;
+    }
+    return true;
   }
 
   nonstd::optional<iroha::keypair_t> KeysManagerImpl::loadKeys() {
     // Try to load from local file
     std::ifstream priv_file(account_name_ + ".priv");
     std::ifstream pub_file(account_name_ + ".pub");
-    if (not priv_file || not pub_file) {
+    if (not priv_file) {
+      log_->error("Could not read '" + account_name_ + ".priv'");
```

Comment:
Imo 'can' is better to use here

Reply:
Also you can add bindings for filename because it's reused

Reply:
I added bindings but `loadFile()` and `deserializeKeypairField()` returns lamda-function and something like lazy-evaluation happens.
```cpp
    auto loadFile = [this](auto filename, auto &res) {
      return [ that = this, &filename, &res ](auto keypair)
          ->nonstd::optional<iroha::keypair_t> { ...  };
    };
    std::string client_pub_key;
    std::string client_priv_key;
    return nonstd::make_optional<iroha::keypair_t>()
               | loadFile(account_name_ + ".pub", client_pub_key)
               | loadFile(account_name_ + ".priv", client_priv_key);
               | deserializeKeypairField(&iroha::keypair_t::pubkey,
                                         client_pub_key)
               | deserializeKeypairField(&iroha::keypair_t::privkey,
                                         client_priv_key)
               | [this](auto keypair) -> nonstd::optional<iroha::keypair_t> {
      return this->validate(keypair) ? nonstd::make_optional(keypair)
                                     : nonstd::nullopt;
```
In the above code, `client_pub_key` and `client_priv_key` are empty and `deserializeKeypairField()` fails.

I changed to:
```cpp
    //TODO 15/11/17 motxx - better way to suppress the evaluation?
    auto blocking = nonstd::make_optional<iroha::keypair_t>()
        | loadFile(account_name_ + ".pub", client_pub_key)
        | loadFile(account_name_ + ".priv", client_priv_key);

    return blocking
               | deserializeKeypairField(&iroha::keypair_t::pubkey,
                                         client_pub_key)
               | deserializeKeypairField(&iroha::keypair_t::privkey,
                                         client_priv_key)
               | [this](auto keypair) -> nonstd::optional<iroha::keypair_t> {
      return this->validate(keypair) ? nonstd::make_optional(keypair)
                                     : nonstd::nullopt;
```
But the code is dirty.